iree/compiler/Dialect/Flow/Transforms/OutlineLargeConstantsPass.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <utility>
+
+#include "iree/compiler/Dialect/Flow/IR/FlowDialect.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+// NOTE: a total guess :) this feels like about the most per-dispatch-buffer
+// data we'd want to embed in the command buffer.
+// TODO(benvanik): make a pass option so users can override.
```

Comment:
The nice thing about MLIR is that we have a way to propagate information from the frontend all the way through to here (attributes). So the tf saved model importer, for example, could read random junk from the protobuf and tag it on the variables as it imported and we could read them here (the saved model importer is the thing that creates the `flow.variable` ops this works on). However in your frontend of choice you'd say something like "this initial variable value is sourced from this byte range of a file" is then an import-time mapping, and (I think I mention it in some comment in this PR) we can have an `#iree.external_constant<filename, offset, length, ...>` attribute that captures that information. That way we can still do things like have a special pass that runs through uses of a value and reads/folds/writes back to files to fold in math or dedupe. The reason I added the `hal.constant_pool.splat` op in this PR is to verify that we could support constants with no storage and just runtime materialization (in the case of splat, alloc+fill, but could be "ask runtime to map this file path" for a hal.constant_pool.external_value or something). The splats aren't really used in real models (we fold those away before we get to outlining constants) but I was able to design everything to work with the more general set because of them (and writing tests is easier!).

Reply:
:D nice, giving hints or explicitly controlling behavior from the frontend for some of these features sounds really useful.